src/box/lua/config/configdata.lua

```diff
@@ -246,6 +253,167 @@ local function effective_snapshot_dir(iconfig)
     return fio.abspath(fio.pathjoin(work_dir, snap_dir))
 end
 
+-- Read snap file and return a map of saved UUIDs and names for
+-- all instances and for the current repicaset.
+local function snapshot_names(snap_path)
+    local peers = {}
+    local instance_uuid = xlog.meta(snap_path).instance_uuid
+    local instance_name, replicaset_name, replicaset_uuid
+    for _, row in xlog.pairs(snap_path) do
+        local body = row.BODY
+        if not body.space_id then
+            goto continue
+        end
+
+        if body.space_id > box.schema.CLUSTER_ID then
+            -- No sense in scanning after _cluster.
+            break
+        end
+
+        if body.space_id == box.schema.SCHEMA_ID then
+            if body.tuple[1] == 'replicaset_uuid' then
+                replicaset_uuid = body.tuple[2]
+            elseif body.tuple[1] == 'replicaset_name' then
+                replicaset_name = body.tuple[2]
+            end
+        elseif body.space_id == box.schema.CLUSTER_ID then
+            if body.tuple[2] == instance_uuid then
+                instance_name = body.tuple[3]
+            end
+
+            table.insert(peers, {
+                instance_uuid = body.tuple[2],
+                instance_name = body.tuple[3],
+            })
+        end
+        ::continue::
+    end
+
+    return {
+        replicaset_name = replicaset_name,
+        replicaset_uuid = replicaset_uuid,
+        instance_name = instance_name,
+        instance_uuid = instance_uuid,
+        peers = peers,
+    }
+end
+
+local function find_peer_name_by_uuid(peers, instance_uuid)
+    for name, peer in pairs(peers) do
+        local uuid = instance_config:get(peer.iconfig_def,
+            'database.instance_uuid')
+        if uuid == instance_uuid then
+            return name
+        end
+    end
+    return nil
+end
+
+-- Validate UUIDs and names passed to config against the data,
+-- saved inside snapshot. Fail early and don't recover all
+-- data if mismatch is found. Returns a map of names, missing
+-- in snapshot in the following format: {<name> = <uuid>}
+local function validate_names(snap_names, config_names)
```

Comment:
Does it allow to start/join an anonymous replica?

Reply:
It does (commit: ` config: validate names and UUIDs before box.cfg `):

```lua
    local iconfig = config_names.peers[config_names.instance_name].iconfig_def
    if snap_names.instance_name == nil and
       config_names.instance_uuid == nil and
       not instance_config:get(iconfig, 'replication.anon') then
        error(string.format('Instance name is not set in snapshot and UUID ' ..
                            'is missing in the config. Found %s in snapshot.',
                            snap_names.instance_uuid), 0)
    end
```

Though it's not tested, as it's impossible to drop UUIDs.